### ISSUE TITLE
Look into porting this to Firefox

### ISSUE THREAD

**Issue by infokiller** [infokiller/google-search-navigator#13]

**Comment by yfdyh000** [comment]
After adding the https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/applications, looks works.

**Comment by Cimbali** [comment]
Well I tried, even with adding `applications`  to the manifest, doesn't seem to work out of the box.

**Comment by yfdyh000** [comment]
It works for me. Check errors and warnings in [browser console](https://developer.mozilla.org/en-US/docs/Tools/Browser_Console)?

**Comment by Cimbali** [comment]
Yeah it's not the most explicit error message. On loading a google page (after having loaded the plugin) the console reports:

> 20:07:44.425 uncaught exception: undefined 1 (unknown)
> 20:07:45.567 window.controllers is deprecated. Do not use it for UA detection. 1 search

I'm running Firefox ESR 52.2.1 (64-bit) on linux (openSUSE).

**Comment by mancze** [comment]
I can't live without this extension and there is no way I'm migrating to Chrome. Hence I'm willing to make this work in Firefox. 

I have added `applications` key to the `manifest.json` file as suggested by yfdyh000 and loaded the temporary extension via `web-ext` npm module. The extension works for me. Tested on version 55.0.3 (64-bit) on Windows 10.

The extension needs its id in order to use `chrome.storage.sync` API. Using `applications` key in the manifest file is one of the option. One might provide id also when signing the assembly `web-ext build --id <id>`. But in such case the debugging (using temporary extension) would be difficult and also `web-ext lint` would trigger warnings about this.

**Comment by mancze** [comment]
So it seems there are two options:

*1) use `applications` key*

In such case such manifest file would be incompatible (or would trigger warning to be more precise) with Chrome. So it would be good idea to have either two manifest files and keep them in sync (duplicity - bad) or have a script which would prepare the sources for the Firefox (better). 

The script would either create a separate copy of the source files under non-versioned directory or modify the manifest directly. I think the former is better. Modifying the versioned file exposes developer to a risk of accidentally committing it which is undesired. In addition this modification would clutter the `git status` output during the development/debugging. Hence I would prefer the first option.

The script could watch for the file changes as the `web-ext` does so the extension reload mechanism would remain working even for Firefox development/testing.

But generally I don't like the idea because it sounds overly complicated at the time being.

*2) sign with `--id` switch*

Instead we can have one manifest file both for Chrome and Firefox. Sources can be modified so that option sync is voluntary and extension works even on sync error. In such case sync couldn't be tested under Firefox prior package signing. But I think we can live with it.


Actually I have already prepared the variant 2: [mancze:firefox-porting](/mancze/google-search-navigator/tree/firefox-porting). The only thing left is testing the signed file with the `--id` switch. The MDN documentation is inconsistent here and I'm not sure it will actually work. :)

**Comment by infokiller** [comment]
Thanks a lot for the detailed writeup @mancze !
Sounds like you can indeed proceed with (2), and I'm of course open to merge these changes to this repo.

**Comment by mancze** [comment]
So I made some progress. I have built and signed the add-on and added it to my Firefox installation.

For this purpose I have prepared build script anyway. I found out that building the extension via `web-ext` ignores some files but not all. Some clutter and especially the `demo.gif` (huge) was added to the resulting package. The command does not accept any kind of ignore list so "clean package sources" directory is needed. This might be achieved via either reorganizing source code (for example introducing `src` directory as is common) or - as I did it - preparing such directory somewhere.

The `web-ext sign` command produces a signed `xpi` add-on which might be distributed. In addition the extension will automatically appear on the [AMO](https://addons.mozilla.org/) portal and you can manage it from there. When `--id <uuid>` parameter is provided, given id is used as add-on id. If no parameter is given, the id will be automatically assigned (I guess). The file `.web-extension-id` is created automatically and contains the add-on id used during signing. Such file will be checked next time signing is run without `--id` parameter. This way the extension will always have same id even when not specified in the `manifest.json`. 

I think the `.web-extension-id` file should be versioned so that add-on id is in the repository. If some individual developer would like to build on their own, they could use parameter `--id` explicitly to override the default id. (The add-on id is linked to the original signer and nobody else can use it.)

**Comment by mancze** [comment]
Hey, I noted that the extension is still not published in Firefox add-on portal. I think that it is pity as the extension should be now fully compatible. @infokiller If you don't mind I will put myself into role of the curator of the extension under AMO portal so you don't need to worry about it. Is that ok with you?

**Comment by infokiller** [comment]
Sorry about that!
Sure, please go ahead and upload it to AMO under whatever name you want.

**Comment by mancze** [comment]
Tada: https://addons.mozilla.org/cs/firefox/addon/the-google-search-navigator/ (url strip without "the" is already taken, probably from some previous attempts). 

As a support e-mail I have used the same which is listed on the Chrome Web Store. Is it ok?

Also I have released it as a version 0.2.7. It is already different from 0.2.5 published in Chrome Web Store so that's the main reason. The reason why it is not 0.2.6 is that the file got somehow broken for download (404).
